### ISSUE TITLE
1.5.0-17: Air Guardian Locks Up The Game + Malfunctioning AI

### ISSUE THREAD

**Issue by ProfHugo** [Growlith1223/ArsMagica2#181]
idk why #37 was closed, but as of now, the air guardian still locks up game. However, now instead of locking up the second it stops spinning, in survival, it would now stand still in its spot, and keeps spinning, over and over again. It now locks up under a number of occasion, such as: changing to creative mode, using /kill or /kill @e, hitting the guardian (for some reason, nothing can hit it, but in the offchance that it did get hit, game locks up), leaving the area where it is at, and many, many more things that could make the game go boom. After the game is borked, and killed, when logged back in, I'm back in the chamber no matter how far I go.

gallery of spinning to win:

Hey why aren't you moving?
![2017-02-05_15 46 20](https://cloud.githubusercontent.com/assets/23370942/22631255/c7968300-ebbd-11e6-970d-9a243e6b7056.png)

funny thing on his head everytime he spin
![2017-02-05_15 46 21](https://cloud.githubusercontent.com/assets/23370942/22631258/e240e4c0-ebbd-11e6-9257-9b4ff571d740.png)

a shot without anything but the guardian rendered
![2017-02-05_16 04 35](https://cloud.githubusercontent.com/assets/23370942/22631261/f31e2cc6-ebbd-11e6-8cef-cac9b049f4d9.png)

teleported out of the area, went pretty far away, can still see the pod (game already borked)
![2017-02-05_16 06 29](https://cloud.githubusercontent.com/assets/23370942/22631268/20c49b06-ebbe-11e6-8f3c-364c9b9555e0.png)

walked a little farther, game didn't load a single chunk, welp
![2017-02-05_16 07 16](https://cloud.githubusercontent.com/assets/23370942/22631274/38b9e6b2-ebbe-11e6-8d2a-2813b38b5641.png)

killed the game there, relogged, now I'm back in the pod with Mr. Funnyhead
![2017-02-05_16 09 54](https://cloud.githubusercontent.com/assets/23370942/22631277/4692eb76-ebbe-11e6-8d35-9e5895dffef2.png)

**Comment by Growlith1223** [comment]
yea im not sure why it was closed either but w/e

it's been a while since anyone has reported this honestly lol

**Comment by ProfHugo** [comment]
I assumed that temporary fix did it and never bothered to check. Welp apparently not, still broken

**Comment by Growlith1223** [comment]
wish i knew what method was locking everything up... can't remember the program i used to use :T

**Comment by Builderb0y** [comment]
Testing it myself, VisualVM says that am2.bosses.ai.EntityAIHurricane.func_75253_b() is what's stalling. Same method as in the last issue.

**Comment by ProfHugo** [comment]
Yeah it's something related to the hurricane attack. I disabled the initiation of that function in the code, with only the other one running (well along with the usual roaming and anti-cc ai), and the air guardian no longer locks up the client (since it can't do the attack anymore). However, it looks like the other function (spawning hazardous roaming tornados that does damage and knocks you up if you touch it) also doesn't work. In short, it is something in the hurricane attack that borks the client, dunno which part tho

**Comment by Growlith1223** [comment]
so, just a quick question, you're able to reproduce it very easily?
EDIT: thanks to decompiling and the decompiler not knowing the function name, i now know what function it is but i need to know if this is easily reproducible lol

**Comment by ProfHugo** [comment]
The lock itself was very easy to reproduce, yes

**Comment by Growlith1223** [comment]
k, cuz im not able to repro this lol, i've had it happen before but doing what you've said has had no effect whatsoever

**Comment by Growlith1223** [comment]
im going to do a new commit, see if BEFORE this new commit it still happens then see if this next commit it happens again

**Comment by ProfHugo** [comment]
Note: I went into the air guardian code and literally // the task initiation of its hurricane attack and recompiled it, thus it no longer locks the game

**Comment by Growlith1223** [comment]
Try building the mod now and using that, see if it no longer locks up

**Comment by Growlith1223** [comment]
also, while you're at it, could you export a snapshot of the cpu info from jvisualVM, might help to know what exactly is going on lol

**Comment by ProfHugo** [comment]
The game still locks up if I: Enters creative mode from survival when the guardian has aggro on me, hit it, etc (same conditions). Also, forgot to note that the second the game locks up, a sound of breaking blocks could be heard before the game is locked. In short, no difference.

EDIT: Will do

**Comment by Growlith1223** [comment]
hm, that's unfortunate, well if you can, post any details you can from jvisualVM, i haven't been able to reproduce this yet, hence why im asking you to do so

**Comment by ProfHugo** [comment]
Some screenshots I gathered up:

1. Game didn't break yet, air guardian free spinning
![1](https://cloud.githubusercontent.com/assets/23370942/22672706/6ebc62c2-ec8a-11e6-9e54-530a680d2243.png)

2. Game Broken Via hitting
![2](https://cloud.githubusercontent.com/assets/23370942/22672733/9100c986-ec8a-11e6-84c9-2d2421a2c22d.png)

3. Before Breaking with Sampling Data
![4](https://cloud.githubusercontent.com/assets/23370942/22672755/b3d2cf18-ec8a-11e6-8551-f97426aaa555.png)

4. After breaking with sampling data
![5](https://cloud.githubusercontent.com/assets/23370942/22672759/be5d0de0-ec8a-11e6-9145-fc6326979990.png)

5. AM2 sampling data right after breaking
![6](https://cloud.githubusercontent.com/assets/23370942/22672778/d7d6b866-ec8a-11e6-8f6e-a2e4a5601b7b.png)

6. Monitor after broken for a while
![7](https://cloud.githubusercontent.com/assets/23370942/22672781/dba6c184-ec8a-11e6-90f4-f734939dd982.png)

7. AM2 data after broken for a while
![8](https://cloud.githubusercontent.com/assets/23370942/22672812/1b88deae-ec8b-11e6-8706-599294f4bba7.png)

8. AM2 data after breaking for a while (full page)
![9](https://cloud.githubusercontent.com/assets/23370942/22672825/2a75b702-ec8b-11e6-8738-d4e719d19624.png)

Notes: The total time of function 70071 and 75235 (hurricane) keeps on adding up and never drops

**Comment by ProfHugo** [comment]
Also, I reproduced this in single player, and to trigger any locking, you have to be "in combat" with the boss in survival mode so that it has aggro on you and will attempt to use hurricane

EDIT: Yep, the boss locks the game if the play attempt to drop aggro, which explains why moving out of the area or going into creative could trigger the lock....

**Comment by Growlith1223** [comment]
try now? if you're still here i mean lol

**Comment by ProfHugo** [comment]
It looks like it no longer gets stuck in an infinite loop now (no longer locks up the game now 😺 ). But the ai is still pretty funky, as it still keeps on doing the attack, over and over again without moving, and spawning whirlwind traps still doesn't work. 

It is a huge step in the right direction I suppose...

**Comment by Growlith1223** [comment]
well, it's a start, at least it doesn't lock up anything anymore... lol as for the continuous attacks, i think i know the reason for it but im going to be heading to sleep so it'll have to wait till tomorrow
